framework/source/class/qx/core/MObjectId.js

```diff
@@ -0,0 +1,234 @@
+     */
+    getObject: function(id) {
+      if (this.__ownedObjects) {
+        var obj = this.__ownedObjects[id];
+        if (obj !== undefined) {
+          return obj;
+        }
+      }
+      
+      // Handle paths
+      if (id.indexOf('/') > -1) {
+        var segs = id.split('/');
+        var target = this;
+        var found = segs.every(function(seg) {
+          if (!seg.length) {
+            return true;
+          }
+          if (!target) {
+            return false;
+          }
+          var tmp = target.getObject(seg);
+          if (tmp !== undefined) {
+            target = tmp;
+            return true;
+          }
+        });
+        return found ? target : undefined;
+      }
+      
+      // No object, creating the object
+      var obj = this._createObjectImpl(id);
+      if (obj !== undefined) {
+        this.addOwnedObject(obj, id);
+      }
+      
+      return obj;
+    },
+    
+    /**
+     * Creates the object, intended to be overridden. Null is a valid return
+     * value and will be cached by `getObject`, however `undefined` is NOT a
+     * valid value and so will not be cached meaning that `_createObjectImpl`
+     * will be called multiple times until a valid value is returned.
+     * 
+     * @param id {String} ID of the object
+     * @return {qx.core.Object?} the created object
+     */
+    _createObjectImpl: function(id) {
+      return;  // Return undefined
+    },
+    
+    /**
+     * Adds an object as owned by this object
+     * 
+     * @param obj {qx.core.Object} the object to register
+     * @param id {String?} the id to set when registering the object
+     */
+    addOwnedObject: function(obj, id) {
+      if (!this.__ownedObjects) {
+        this.__ownedObjects = {};
+      }
+      var thatOwner = obj.getOwner();
+      if (thatOwner === this) {
+        return;
+      }
+      if (thatOwner) {
+        thatOwner.removeOwnedObject(obj);
+      }
+      if (id === undefined) {
+        id = obj.getObjectId();
+      }
+      if (!id) {
+        throw new Error("Cannot register an object that has no ID, obj=" + obj);
+      }
+      if (this.__ownedObjects[id]) {
+        throw new Error("Cannot register an object with ID '" + id + "' because that ID is already in use, this=" + this + ", obj=" + obj);
+      }
+      if (obj.getOwner() != null) {
+        throw new Error("Cannot register an object with ID '" + id + "' because it is already owned by another object this=" + this + ", obj=" + obj);
+      }
+      obj.__changingOwner = true;
+      try {
+        obj.setOwner(this);
+        obj.setObjectId(id);
+      } finally {
+        obj.__changingOwner = false;
+      }
+      this.__ownedObjects[id] = obj;
+    },
+
+    /**
+     * Discards an object from the list of owned objects; note that this does
+     * not dispose of the object, simply forgets it if it exists.
+     * 
+     * @param id {String|Object} the ID of the object to discard, or the object itself
```

Comment:
Calling an object `id` is a bit confusing. How about calling this variable `data` and then set local variables `id` and `obj` appropriately?